src/checkers/inference/model/ImplicationConstraint.java

```diff
@@ -0,0 +1,110 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Constraint that models implication logic. If all the assumptions are satisfied, then
+ * conclusion should also be satisfied.
+ *
+ * This is intended to express the restrictions between solutions for {@link Slot}s.
+ *
+ * Suppose one needs to enforce this restriction(just an example. Doesn't indicate universal type rule):
+ * if {@code @A} is inferred as on declaration of class {@code MyClass}, then every usage of class
+ * {@code MyClass} needs to be inferred to {@code @A}, but to nothing else. {@link ImplicationConstraint}
+ * can express this restriction:
+ * <p>
+ * {@code @1 == @A -> @2 == @A}, in which {@code @1} is the slot inserted on the class tree and
+ * {@code @2} is the slot that represents one usage of {@code MyClass}.
+ */
+public class ImplicationConstraint extends Constraint {
+
+    /**
+     * A list of {@link Constraint}s that are conjuncted together.
+     */
+    private final List<Constraint> assumptions;
+
+    /**
+     * A single {@link Constraint} that is implicated by the {@link #assumptions}.
+     */
+    private final Constraint conclusion;
+
+    public ImplicationConstraint(
+            List<Constraint> assumptions, Constraint conclusion, AnnotationLocation location) {
+        super(extractAllSlots(assumptions, conclusion), location);
+
+        this.assumptions = assumptions;
+        this.conclusion = conclusion;
+    }
+
+    private static List<Slot> extractAllSlots(List<Constraint> assumptions, Constraint conclusion) {
+        List<Slot> slots = new ArrayList<>();
+        for(Constraint a : assumptions) {
+            slots.addAll(a.getSlots());
+        }
+        slots.addAll(conclusion.getSlots());
+        return slots;
+    }
+
+    public List<Constraint> getAssumptions() {
+        return assumptions;
+    }
+
+    public Constraint getConclusion() {
+        return conclusion;
+    }
+
+    @Override
+    public <S, T> T serialize(Serializer<S, T> serializer) {
+        return serializer.serialize(this);
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        for(Constraint a : assumptions) {
+            result = result * prime + a.hashCode();
+        }
+        result = result * prime + conclusion.hashCode();
+        return result;
+    }
+
+    public static Constraint create(List<Constraint> assumptions, Constraint conclusion, AnnotationLocation currentLocation) {
+        if (assumptions == null || conclusion == null) {
+            throw new BugInCF("Adding implication constraint with null argument. assumptions: "
+                    + assumptions + " conclusion: " + conclusion);
+        }
+
+        if (assumptions.isEmpty()) {
+            // Optimization for trivial cases when there are no preconditions for the conclusion to be true
+            // , meaning conclusion is a hard Constraint that must be satisfied.
+            return conclusion;
+        }
+
```

Comment:
the normalization cases I see are:

1) assumptions == empty || refinedAssumptions == empty ==> return conclusion
2) any assumption == false ==> return True
3) refinedAssumptions != empty && conclusion == true ==> return True
4) refinedAssumptions != empty && conclusion == false ==> return conjunction of refinedAssumptions
5) refinedAssumptions != empty && conclusion != true && conclusion != false ==> return a real implication constraint

for 3. if refinedAssumptions are not empty but the conclusion is true, the whole thing can also normalized to true; ie "A implies True":
if A is true, the whole expression is true.
if A is false, the whole expression is also true.

for 4. "not A or False" can only be satisfied by getting sat in "not A" so the "or False" doesn't have to be encoded. I remember that we had discussed about creating a "ConjunctionConstraint" for this at some point but it wasn't going to be with this PR. Can you add a TODO here to use the ConjunctionConstraint in the future for this normalization case?

Please add code for case 3.